### ISSUE TITLE
[4.0] Repeatable Modal fields and subforms

### ISSUE THREAD

**Issue by brianteeman** [joomla/joomla-cms#32663]
This is **not** about the user or media fields as they are different and there already is a PR to fix those.

This is about the other modal fields such as the article tester

### Steps to reproduce the issue
Edit `administrator\components\com_users\forms\user.xml` by inserting the following code at the beginning of the file after the opening `form` ie on line4

```
		<fieldset name="testing" addfieldprefix="Joomla\Component\Content\Administrator\Field">
			<field name="field-name" type="subform" label="SUBFORM" multiple="true">
				<form>
					<field name="name" type="modal_article" label="ARTICLE"/>
				</form>
			</field>
		</fieldset>
```

If you try to add a new user you will see a new tab as shown in the image
![image](https://user-images.githubusercontent.com/1296369/111027908-0dfa6900-83eb-11eb-8772-0edd4ca9900f.png)

Click on the Select and add an article
Click on the green plus to repeat the field 
Click on the Select and you cannot add an article as the modal is empty

Basically if the field is on the page when it loads then the modal works. If not then it is empty

**Comment by dgrammatiko** [comment]
> Basically if the field is on the page when it loads then the modal works. If not then it is empty

That was the reason I kept saying that all fields need to be implemented as custom elements (or web components). Custom elements have lifecycle events, so if you create (as it happens with subform) a new field by copying the html then the field initialises automatically (the lifecycle event for this is `connectedCallback`) 

Can it be done without custom elements? Yes but you need to set up event listeners and it would get messy. Not to mention that having multiple ways for the same task is extremely confusing for devs (when should I use this method over the other, etc)

Unfortunately the last 1-2 years nothing happen to bring the remaining fields to parity with the existing ones and on top of that people added interactivity (or new fields) that are totally incompatible with subforms (the color slider, the text count characters, etc). I had a list https://github.com/joomla/joomla-cms/issues/19427 which apparently was never implemented although the code was written...

**Comment by Fedik** [comment]
Here multiple issues:

* initialization (easy)
* hardcoded IDs (hard)
* bunch of callbacks (eg window.processModalX(), ) with hardcoded ID parameters (much hard)
* bootstrap modal (hard)

Custom elements can help only with initialization.
The problem much bigger than that.

@brianteeman unfortunately it cannot be fixed, without a loot of rewriting. 
For now I can suggest a work around:
* Add row with a field
* Save
* Select value

This somehow related to what I tries to explain in #32473 discussion.

**Comment by brianteeman** [comment]
Sorry I dont understand your work around.

For my own component I am just forking the users webcomponent

**Comment by Fedik** [comment]
I will try differently:
The site User (eg:Admin, Manager) should add a new row in the subform field, without selecting anything,
Then save the content item, 
And only then select the value for this added field.

**Comment by brianteeman** [comment]
I realised what you meant as soon as I hit comment ;)

**Comment by dgrammatiko** [comment]
> bunch of callbacks (eg window.processModalX(), ) with hardcoded ID parameters (much hard)

@Fedik this is the worse part and unfortunately cannot be done in any B/C way. But B/C might be ignored here (the modal fields or whatever the name) is an internal thing. There was a PR that was solving aspects of this but not the callbacks

**Comment by Fedik** [comment]
This will be fixed in Joomla 5, in
- #40462

For now only for `modal_article`, the rest of modal fields still need to be converted.
You can test after beta1 release, or on current 5.0-dev branch.

**Comment by Fedik** [comment]
Most of the fields is fixed:
- https://github.com/joomla/joomla-cms/pull/40462
- https://github.com/joomla/joomla-cms/pull/41629
- https://github.com/joomla/joomla-cms/pull/41632
- https://github.com/joomla/joomla-cms/pull/41680
- https://github.com/joomla/joomla-cms/pull/42293
- https://github.com/joomla/joomla-cms/pull/42326
- https://github.com/joomla/joomla-cms/pull/42327
- https://github.com/joomla/joomla-cms/pull/42288

Some still need to be tested.
If anything is missing then please open new issue.
